opm/simulators/linalg/DILU.hpp

```diff
@@ -0,0 +1,183 @@
+#include <opm/simulators/linalg/PreconditionerWithUpdate.hpp>
+
+#include <dune/common/fmatrix.hh>
+#include <dune/common/version.hh>
+#include <dune/common/unused.hh>
+#include <dune/istl/bcrsmatrix.hh>
+
+
+namespace Dune
+{
+
+/*! \brief The sequential DILU preconditioner.
+
+   \tparam M The matrix type to operate on
+   \tparam X Type of the update
+   \tparam Y Type of the defect
+*/
+template <class M, class X, class Y>
+class SeqDilu : public PreconditionerWithUpdate<X, Y>
+{
+    public:
+        //! \brief The matrix type the preconditioner is for.
+        using matrix_type = M;
+        //! \brief The domain type of the preconditioner.
+        using domain_type = X;
+        //! \brief The range type of the preconditioner.
+        using range_type = Y;
+        //! \brief The field type of the preconditioner.
+        using field_type = typename X::field_type;
+        //! \brief scalar type underlying the field_type
+
+    /*! \brief Constructor.
+       Constructor gets all parameters to operate the prec.
+       \param A The matrix to operate on.
+    */
+    SeqDilu(const M& A)
+        : A_(A)
+    {
+        Dinv_.resize(A_.N());
+        // the Dinv matrix must be initialised
+        update();
+    }
+
+    /*!
+       \brief Update the preconditioner.
+       \copydoc Preconditioner::update()
+    */
+    virtual void update() override
+    {
+        OPM_TIMEBLOCK(update);
+        
+        auto endi = A_.end();
+        for ( auto row_i = A_.begin(); row_i != endi; ++row_i) {
+            Dinv_[row_i.index()] = A_[row_i.index()][row_i.index()];
+        }
+        
+        for ( auto row_i = A_.begin(); row_i != endi; ++row_i)
+        {
+           auto Dinv_temp = Dinv_[row_i.index()];
+            for (auto a_ij = row_i->begin(); a_ij.index() < row_i.index(); ++a_ij) {
+                auto a_ji = A_[a_ij.index()].find(row_i.index());
+                // if A[i, j] != 0 and A[j, i] != 0
+                if (a_ji != A_[a_ij.index()].end()) {
+                    // Dinv_temp -= A[i, j] * d[j] * A[j, i]
+                    Dinv_temp -= (*a_ij) * Dune::FieldMatrix(Dinv_[a_ij.index()]) * (*a_ji);
+                }
+            }
+            Dinv_temp.invert();
+            Dinv_[row_i.index()] = Dinv_temp;
+        }
+    }
+
+    /*!
+       \brief Prepare the preconditioner.
+       \copydoc Preconditioner::pre(X&,Y&)
+    */
+    virtual void pre(X& v, Y& d) override
+    {
+        DUNE_UNUSED_PARAMETER(v);
+        DUNE_UNUSED_PARAMETER(d);
+    }
+
+
+    /*!
+       \brief Apply the preconditioner.
+       \copydoc Preconditioner::apply(X&,const Y&)
+    */
+    virtual void apply(X& v, const Y& d) override
+    {
+        // M = (D + L_A) D^-1 (D + U_A)   (a LU decomposition of M)
+        // where L_A and U_A are the strictly lower and upper parts of A and M has the properties:
+        // diag(A) = diag(M)
+        // solving the product M^-1(b-Ax) using upper and lower triangular solve
```

Comment:
Referring to `b - Ax` here, maybe rewrite slightly the comments referring to `d` and `v` instead?

Reply:
I was a bit confused with the use of v,d and b,x in OPM/Dune initially, but now that the algorithm is implemented to work with the defect `d` and update `v` it makes sense to change the comments.